### PR TITLE
test: pod distribution on Multi-AZ cluster

### PR DESCRIPTION
**Pull request by psturc**
# Description
https://issues.redhat.com/browse/MGDAPI-59

# Comments
* this test verifies the correct pod distribution of 3scale, User SSO and RHSSO services (by default)
* it allows to test a pod distribution in a custom namespace (by passing in an env var with the list of namespaces)

# Verification steps
1. Clone this branch
2. Target multi-az cluster with managed-api installed (ping me if you need such cluster)

### Verify: Tests passes

1. Run
```
go clean -testcache && MULTIAZ=true go test -v ./test/functional -run="//^F09"
```
2. Verify that the test has passed (see something like `--- PASS: TestIntegreatly/API_Managed_Multi-AZ_Tests/F09_-_Verify_correct_pod_distribution_on_Multi-AZ_cluster (3.16s)` in the log)

### Verify: Test wasn't run

1. Run
```
go clean -testcache && go test -v ./test/functional -run="//^F09"
```
2. Verify that the test wasn't run because the env var `MULTIAZ` was't defined (see `--- SKIP: TestIntegreatly/API_Managed_Multi-AZ_Tests (0.00s)` in the log)

### Verify: `NAMESPACES_TO_CHECK` env var works

1. Uncomment [this line](https://github.com/integr8ly/integreatly-operator/pull/1201/files#diff-e7a8d8d902bb6b36f0482f7895b49926R145) to log the detailed information about the pod distribution per pods' owner
2. Run
```
go clean -testcache && MULTIAZ=true NAMESPACES_TO_CHECK=redhat-rhmi-operator,redhat-rhmi-3scale-operator go test -v ./test/functional -run="//^F09"
```
3. You should see something like this in the test output log
```
Pods distribution in rhmi-operator-688f6fdccb: &{zones:map[eu-central-1a:1] podsTotal:1 namespace:redhat-rhmi-operator}
Pods distribution in rhmi-registry-cs: &{zones:map[eu-central-1c:1] podsTotal:1 namespace:redhat-rhmi-3scale-operator}
```

### Verify: Pod distribution check works, even when an invalid namespace is passed

1. Run
```
go clean -testcache && MULTIAZ=true NAMESPACES_TO_CHECK=invalid-namespace,redhat-rhmi-operator go test -v ./test/functional -run="//^F09"
```
2. You should see an error in the log mentioning the "invalid-namespace", but there should be also a log mentioning the pod distribution in redhat-rhmi-operator

### Verify: test works after changing the amount of replicas for a service
1. Scale down 3scale operator and scale up apicast-production
```
oc scale deployment --replicas 0 3scale-operator -n redhat-rhmi-3scale-operator
oc scale dc --replicas 5 apicast-production -n redhat-rhmi-3scale
```
2. Wait until all pods belonging to apicast-production dc are running
```
watch oc get pods -n redhat-rhmi-3scale -l deploymentConfig=apicast-production
```
3. Run the test
```
go clean -testcache && MULTIAZ=true go test -v ./test/functional -run="//^F09"
```
4. You should see something like this in the test output
```
Pods distribution in apicast-production-1: &{zones:map[eu-central-1a:2 eu-central-1b:1 eu-central-1c:2] podsTotal:5 namespace:redhat-rhmi-3scale}
```
indicating that the total number of pods changed to 5 and they are distributed evenly across the AZs

### CODE REVIEW

**Review by austincunningham — COMMENTED** [review]
/lgtm
will try and verify today

**Comment by austincunningham** [comment]
I am seeing failures but I have been killing zones so don't expect the pods to be evenly distributed 
```
=== RUN   TestIntegreatly/API_Managed_Multi-AZ_Tests/F09_-_Verify_correct_pod_distribution_on_Multi-AZ_cluster
    TestIntegreatly/API_Managed_Multi-AZ_Tests/F09_-_Verify_correct_pod_distribution_on_Multi-AZ_cluster: multiaz_pod_distribution.go:83: 
        Error when verifying the pod distribution: 
        [Pod owner 'backend-worker-1' pods are not distributed correctly. map[apicast-production-1:0xc000203dc0 apicast-staging-1:0xc000203e00 backend-cron-1:0xc000203e40 backend-listener-1:0xc000203e80 backend-worker-1:0xc000203f00 system-app-2:0xc000203f60 system-memcache-1:0xc000203f80 system-sidekiq-2:0xc000203fa0 system-sphinx-1:0xc000203fc0 zync-1:0xc000203fe0 zync-database-1:0xc000c96040 zync-que-1:0xc000c96100]
         Pod owner 'system-sidekiq-2' pods are not distributed correctly. map[apicast-production-1:0xc000203dc0 apicast-staging-1:0xc000203e00 backend-cron-1:0xc000203e40 backend-listener-1:0xc000203e80 backend-worker-1:0xc000203f00 system-app-2:0xc000203f60 system-memcache-1:0xc000203f80 system-sidekiq-2:0xc000203fa0 system-sphinx-1:0xc000203fc0 zync-1:0xc000203fe0 zync-database-1:0xc000c96040 zync-que-1:0xc000c96100]
         Pod owner 'system-app-2' pods are not distributed correctly. map[apicast-production-1:0xc000203dc0 apicast-staging-1:0xc000203e00 backend-cron-1:0xc000203e40 backend-listener-1:0xc000203e80 backend-worker-1:0xc000203f00 system-app-2:0xc000203f60 system-memcache-1:0xc000203f80 system-sidekiq-2:0xc000203fa0 system-sphinx-1:0xc000203fc0 zync-1:0xc000203fe0 zync-database-1:0xc000c96040 zync-que-1:0xc000c96100]
        ]
```
going to try a fresh install

File: test/functional/multiaz_pod_distribution.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+func testCorrectPodDistribution(dist map[string]*podDistribution) []string {
+	var testErrors []string
+	for podOwner, pd := range dist {
+		minPodsPerZone, maxPodsPerZone := getAllowedNumberOfPodsPerZone(pd.podsTotal)
+		// fmt.Printf("Pods distribution in %s: %+v\n", podOwner, pd)
+		for _, n := range pd.zones {
+			if n == minPodsPerZone || n == maxPodsPerZone {
+				continue
+			}
+			testErrors = append(testErrors, fmt.Sprintf("Pod owner '%s' pods are not distributed correctly. %+v\n", podOwner, dist))
```
Comment (matskiv) [inline test/functional/multiaz_pod_distribution.go:150]:
```suggestion
			testErrors = append(testErrors, fmt.Sprintf("Pod owner '%s' pods are not distributed correctly. %+v\n", podOwner, pd))
```

File: test/functional/integreatly_test.go [hunk 2]
```diff
@@ -47,4 +47,13 @@ func TestIntegreatly(t *testing.T) {
...
 	})
+
+	t.Run("API Managed Multi-AZ Tests", func(t *testing.T) {
+		// Do not execute these tests unless MULTIAZ is set to true
+		if os.Getenv("MULTIAZ") != "true" {
+			t.Skip("Skipping Multi-AZ tests as MULTIAZ env var is not set to true")
+		}
+
+		common.RunTestCases(MULTIAZ_TESTS, t, config)
+	})
```
Comment (matskiv) [inline test/functional/integreatly_test.go:58]: Running these tests before destructive tests could be safer.
Or is this put here on purpose? :)

**Comment by matskiv** [comment]
Two comments from me above.
Otherwise LGTM, nice code comments 👍

File: test/functional/integreatly_test.go (hunk 2 above)
Comment (psturc) [reply test/functional/integreatly_test.go:58]: No, it is a good idea to run these tests before the destructive ones, I'll add the change.

**Comment by austincunningham** [comment]
fresh install still seeing a failure 
```bash
=== RUN   TestIntegreatly/API_Managed_Multi-AZ_Tests/F09_-_Verify_correct_pod_distribution_on_Multi-AZ_cluster
    TestIntegreatly/API_Managed_Multi-AZ_Tests/F09_-_Verify_correct_pod_distribution_on_Multi-AZ_cluster: multiaz_pod_distribution.go:83: 
        Error when verifying the pod distribution: 
        [Pod owner 'system-sidekiq-2' pods are not distributed correctly. map[apicast-production-1:0xc000712f80 apicast-staging-1:0xc000712fa0 backend-cron-1:0xc000712fc0 backend-listener-1:0xc000712fe0 backend-worker-1:0xc000713000 system-app-1:0xc000713020 system-app-2:0xc000713040 system-memcache-1:0xc000713080 system-sidekiq-2:0xc0007130a0 system-sphinx-1:0xc0007130c0 zync-1:0xc0007130e0 zync-database-1:0xc000713100 zync-que-1:0xc000713120]
         Pod owner 'backend-cron-1' pods are not distributed correctly. map[apicast-production-1:0xc000712f80 apicast-staging-1:0xc000712fa0 backend-cron-1:0xc000712fc0 backend-listener-1:0xc000712fe0 backend-worker-1:0xc000713000 system-app-1:0xc000713020 system-app-2:0xc000713040 system-memcache-1:0xc000713080 system-sidekiq-2:0xc0007130a0 system-sphinx-1:0xc0007130c0 zync-1:0xc0007130e0 zync-database-1:0xc000713100 zync-que-1:0xc000713120]

```
had a look at the pod distribution for my cluster 
```bash
./pods.sh redhat-rhmi-3scale  
Pods distribution for 'redhat-rhmi-3scale'
| Pod name | Availability Zone |
| -------- | ----------------- |
| apicast-production-1-deploy | eu-west-1c |
| apicast-production-1-pg9zl | eu-west-1b |
| apicast-production-1-vgncn | eu-west-1c |
| apicast-staging-1-6xcw6 | eu-west-1c |
| apicast-staging-1-deploy | eu-west-1c |
| apicast-staging-1-szddq | eu-west-1b |
| backend-cron-1-68slw | eu-west-1c |
| backend-cron-1-deploy | eu-west-1a |
| backend-cron-1-q688w | eu-west-1c |
| backend-listener-1-deploy | eu-west-1b |
| backend-listener-1-p6zjr | eu-west-1b |
| backend-listener-1-xrz5w | eu-west-1a |
| backend-worker-1-5sjqt | eu-west-1c |
| backend-worker-1-9jlxp | eu-west-1b |
| backend-worker-1-deploy | eu-west-1b |
| system-app-1-9h64g | eu-west-1b |
| system-app-1-deploy | eu-west-1a |
| system-app-1-hook-post | eu-west-1c |
| system-app-1-hook-pre | eu-west-1b |
| system-app-2-deploy | eu-west-1a |
| system-app-2-gflxj | eu-west-1b |
| system-app-2-hook-pre | eu-west-1c |
| system-app-2-jl2gd | eu-west-1c |
| system-memcache-1-9dqsf | eu-west-1c |
| system-memcache-1-deploy | eu-west-1b |
| system-sidekiq-1-deploy | eu-west-1c |
| system-sidekiq-2-27w64 | eu-west-1c |
| system-sidekiq-2-deploy | eu-west-1c |
| system-sidekiq-2-kxbnk | eu-west-1c |
| system-sphinx-1-5j6b4 | eu-west-1b |
| system-sphinx-1-deploy | eu-west-1a |
| zync-1-22frd | eu-west-1c |
| zync-1-deploy | eu-west-1a |
| zync-1-shbtd | eu-west-1a |
| zync-database-1-deploy | eu-west-1c |
| zync-database-1-qh5s6 | eu-west-1c |
| zync-que-1-cn2pz | eu-west-1a |
| zync-que-1-deploy | eu-west-1c |
| zync-que-1-q5f92 | eu-west-1c |
```

**Comment by psturc** [comment]
@austincunningham I think this might be the problem:
```
...
| backend-cron-1-68slw | eu-west-1c |
| backend-cron-1-deploy | eu-west-1a | # Has status: Completed, does not count
| backend-cron-1-q688w | eu-west-1c |
...
| system-sidekiq-1-deploy | eu-west-1c | # Has status: Completed, does not count
| system-sidekiq-2-27w64 | eu-west-1c |
| system-sidekiq-2-deploy | eu-west-1c | # Has status: Completed, does not count
| system-sidekiq-2-kxbnk | eu-west-1c |
```
it looks like that there are 3 zones available, but these pods are distributed only in zone `eu-west-1c`

**Comment by austincunningham** [comment]
@psturcn I agree but don't think 3scale distribute them by default.

> this comment in jira https://issues.redhat.com/browse/MGDAPI-55?focusedCommentId=15183412&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15183412 suggests that system-sidekiq and backend-cron don't benefit from multi az maybe you can filter your results accordingly

**Comment by austincunningham** [comment]
@psturc this test does what is suppose to do, it just that pods are not distributed across all az at present for our clusters so this functional test will always fail , there are tickets on the backlog to address this https://issues.redhat.com/browse/MGDAPI-64 https://issues.redhat.com/browse/MGDAPI-55 . Not sure if we can merge at this time WDYT?
/lgtm

**Comment by psturc** [comment]
Just realized that I should add the test case for this automated test as well (to describe in details what this test should do), so I will add that to this PR.

But after that I think it is ok to merge it, because by default this test is automatically skipped (it requires a special env var to run), so it shouldn't cause any issues in our pipelines.

Once it's ready to be used on a Multi-AZ cluster, it will be added to the pipeline ([jira ticket](https://issues.redhat.com/browse/MGDAPI-128))

**Comment by austincunningham** [comment]
have use this test many times and its results are consistent with what is found on cluster and am happy to 
/lgtm
/approve

**Pull request merged**